colossalai/quantization/fp8.py

```diff
@@ -0,0 +1,105 @@
+import torch
+import torch.distributed as dist
+
+
+def cast_to_fp8(inp: torch.Tensor, scale=None, fp8_format="e4m3") -> (torch.Tensor, torch.Tensor):
+    r"""
+    casting torch Tensor into specified fp8 tensor with per-channel scaling or per-tensor scaling.
+    Args:
+        inp: input torch Tensor, should be in torch.FloatTensor, torch.HalfTensor, torch.BFloat16Tensor.
+        scale: scaling factor for fp8 casting. If it is None, then it is computed automatically. Per-channel scaling
+        is applied if input tensor is 2 dimension, otherwise, per-tensor scaling is applied.
+        fp8_format: e4m3 or e5m2
+
+    Returns:
+        Tuples: A tuple (fp8_tensor, scale)
+    """
+    if inp.dtype not in [torch.float32, torch.float16, torch.bfloat16]:
+        return inp
+    fp8_type = torch.float8_e4m3fn if fp8_format == "e4m3" else torch.float8_e5m2
+
+    if inp.dim() == 2:
+        if scale is None:
+            per_channel_max = inp.abs().max(dim=-1).values
+            scale = per_channel_max
+        scale_inv = 1.0 / scale
+        scale_inv = scale_inv[:, None]
+        ret = (scale_inv * inp).to(fp8_type)
+    else:
+        if scale is None:
+            per_tensor_max = inp.abs().max()
+            scale = per_tensor_max
+        scale_inv = 1.0 / scale
+        ret = (scale_inv * inp).to(fp8_type)
+
+    return ret, scale
+
+
+def cast_from_fp8(inp: torch.Tensor, scale: torch.Tensor, ret_type: torch.dtype) -> torch.Tensor:
+    r"""
+
+    Args:
+        inp: should be a fp8 torch tensor in one of the types: [torch.float8_e4m3fn, torch.float8_e5m2].
+        scale: scaling factor returned by cast_to_fp8 function.
+        ret_type: the datatype of the returned tensor.
+
+    Returns:
+        torch.Tensor
+    """
+    if inp.dtype not in [torch.float8_e4m3fn, torch.float8_e5m2]:
+        return inp
```

Comment:
Should this raise a warning or an error?

Reply:
Yes, it should raise an error.